### PR TITLE
Add github workflow to upload test failure markdown summary report as a comment to PRs when tests fail

### PR DESCRIPTION
**Pull request by nyalldawson**
Relates to https://github.com/qgis/QGIS-Enhancement-Proposals/issues/268

This PR expands the current mechanism which generates HTML summary reports of test failures to also generate a (concise!) markdown summary report. The markdown report is then added as a comment to the PR, providing a much more submitter-friendly way of determining why a PR is failing the test suites.

A few notes:

- In https://github.com/qgis/QGIS-Enhancement-Proposals/issues/268 I wrote: "The goal is that if a rendering test fails, a comment will be automatically added to the Pull Request containing a descriptive message and a link to download the rendered images for debugging". This is unfortunately still **NOT** possible given the available Github APIs.
- Specifically, I was originally hoping to attach the full test summary artifact (html report + rendered images) as a zip file to the autogenerated comment. However, there's no public API to upload assets to Github. (I did find https://github.com/j178/github-s3, but that relies on a private/non-stable API, and I'm dead against relying on private API in our CI setup... there's enough regular ongoing maintainence required here without intentionally heading into unstable territory :scream_cat: )
- For the same reason, I can't directly embed the rendered images into the auto-created comment -- there's no way we can grab the images from the test report artifact and upload them as issue assets. => The test failure comment is PLAIN TEXT ONLY :disappointed: 
- Failing this, I was hoping to at least put a link in the comment to the test results artifact zip, to make it easy to locate this artifact. But again, there's no Github support for this :scream_cat: :scream_cat: . (See https://github.com/actions/upload-artifact/issues/50 , https://github.com/actions/upload-artifact/issues/53 )
- The comment only relates to test failures relating to rendered image checks -- other general test failures are excluded. The rendered image check is our most unusual/trickiest barrier to new contributors wrt failing tests, so I'm exclusively targeting that here.
- Tests MUST use the base class QgsTest / QgisTestCase methods for render/layout rendering checks. If they have ad-hoc implementations then the markdown report won't include the failures from those ad-hoc methods. Updating all tests to use the base class methods is a WIP, being done on demand (see eg https://github.com/qgis/QGIS/pull/54882/commits/117f5087738c3696372534733c937e8c3649259b )
- In the QEP I've mentioned: "If possible, the workflow improvements will be backported to apply to all maintained stable release branches (including LTR).". This **ISN'T** suitable for backporting... there's been many many changes to the test infrastructure since LTR and the delta is just too large to make a backport workable.

**Also note: this workflow won't actually DO anything from this PR -- it will only kick in when the workflow is present in the master branch.**

### CODE REVIEW

**Comment by nyalldawson** [comment]
Ping @3nids

**Review by nirvn — APPROVED** [review]

**Comment by 3nids** [comment]
Not sure this would work but what about using a dedicated repo and Gihtub pages?

* from the testing workflow, you fork the gh-pages repo, add your test data in a folder named `date-PR-commit`, you can comment with the link
* on the gh-pages repo:
  * an auto-merge workflow which checks the structure of the PR (no file deletion, repsect the folder structure, only html + images, etc.)
  * a cron workflow which automatically deletes the content older than a week or 2

**Review by 3nids — APPROVED** [review]

**Comment by nyalldawson** [comment]
@3nids 

> Not sure this would work but what about using a dedicated repo and Gihtub pages?

Novel approach! My reservation would be that it's adding **considerable** complexity and the accompanying maintenance burden to our already complex CI setup.

My gut feeling is that we should stick within the the more well-trodden, "supported" realm of github workflows. I think there's a good chance that the missing functionality we need is added to github api in future (there's certainly a LOT of demand for it on the various tickets), and at that stage it will be relatively simple to safely upgrade the workflow to handle artifact links/embedded images.

File: .github/workflows/run-tests.yml [hunk 1]
```diff
@@ -416,11 +416,18 @@ jobs:
           mkdir -p /tmp/minio_tests/test-bucket && chmod -R 777 /tmp/minio_tests
           docker-compose -f .docker/$DOCKERFILE run qgis-deps /root/QGIS/.docker/docker-qgis-test.sh $TEST_BATCH
 
+      - name: Save PR number to test report
+        if: ${{ failure() }}
+        env:
+          PR_NUMBER: ${{ github.event.number }}
+        run: |
+          echo $PR_NUMBER > qgis_test_report/pr_number
```
Comment (agiudiceandrea) [inline .github/workflows/run-tests.yml:424]: @nyalldawson, it looks like this command always fails (when there is a test failure) with "permission denied" error:
```
Run echo $PR_NUMBER > qgis_test_report/pr_number
  echo $PR_NUMBER > qgis_test_report/pr_number
  shell: /usr/bin/bash -e {0}
  env:
    DOCKER_USERNAME: 
    DOCKER_PASSWORD: 
    QGIS_WORKSPACE: /home/runner/work/QGIS/QGIS
    CTEST_BUILD_NAME: PR55[2](https://github.com/qgis/QGIS/actions/runs/6843501155/job/18608800408?pr=55264#step:14:2)64_40efa27261fa9511a[3](https://github.com/qgis/QGIS/actions/runs/6843501155/job/18608800408?pr=55264#step:14:3)88730af3bed19563c9682f_ALL_BUT_PROVIDERS
    QT_VERSION: 5
    PR_NUMBER: 5526[4](https://github.com/qgis/QGIS/actions/runs/6843501155/job/18608800408?pr=55264#step:14:4)
/home/runner/work/_temp/f6b1d229-1adc-431a-a7e9-e61[5](https://github.com/qgis/QGIS/actions/runs/6843501155/job/18608800408?pr=55264#step:14:5)d98db7[6](https://github.com/qgis/QGIS/actions/runs/6843501155/job/18608800408?pr=55264#step:14:6)c.sh: line 1: qgis_test_report/pr_number: Permission denied
Error: Process completed with exit code
```

Comment (nyalldawson) [reply .github/workflows/run-tests.yml:424]: @agiudiceandrea I'm working on it - see https://github.com/qgis/QGIS/pull/54919

**Pull request merged**